api/post.go

```diff
@@ -728,6 +730,92 @@ func updateMentionCountAndForget(channelId, userId string) {
 	}()
 }
 
+func checkForOutOfChannelMentions(c *Context, post *model.Post, channel *model.Channel) {
+	// don't check for out of channel mentions in direct channels
+	if channel.Type == model.CHANNEL_DIRECT {
+		return
+	}
+
+	mentioned := getOutOfChannelMentions(post, channel.TeamId)
+	if len(mentioned) == 0 {
+		return
+	}
+
+	usernames := make([]string, len(mentioned))
+	for i, user := range mentioned {
+		usernames[i] = user.Username
+	}
+	sort.Strings(usernames)
+
+	var messageText string
+	if len(usernames) == 1 {
+		messageText = c.T("api.post.check_for_out_of_channel_mentions.message.one", map[string]interface{}{
+			"Username": usernames[0],
+		})
+	} else {
+		messageText = c.T("api.post.check_for_out_of_channel_mentions.message.multiple", map[string]interface{}{
+			"Usernames":    strings.Join(usernames[:len(usernames)-1], ", "),
+			"LastUsername": usernames[len(usernames)-1],
+		})
+	}
+
+	// create an ephemeral post that will be sent only to the sender of this original post and not stored in the DB
+	warningPost := model.Post{
+		Id:        model.NewId(),
+		ChannelId: post.ChannelId,
+		Message:   messageText,
+		Type:      model.POST_OUT_OF_CHANNEL_MENTION,
+		CreateAt:  post.CreateAt + 1,
+		Ephemeral: true,
+		Props:     model.StringInterface{},
+		Filenames: []string{},
+	}
+
+	message := model.NewMessage(c.Session.TeamId, channel.Id, post.UserId, model.ACTION_EPHEMERAL_MESSAGE)
```

Comment:
I was originally using model.ACTION_POSTED for this (the client uses the same handling for it as it does for model.ACTION_EPHEMERAL_MESSAGE), but I needed some way to restrict the message to only be sent to the one user. The web hub code doesn't have a way to look at the contents of the post at that point since it's already been serialized as json.